cachi2/core/models/validators.py

```diff
@@ -24,6 +24,31 @@ def unique(items: Iterable[T], by: Callable[[T], Any], dedupe: bool = True) -> l
     return list(by_key.values())
 
 
+def unique_multikey(items: Iterable[T], by: Callable[[T], List[Any]]) -> list[T]:
+    """Make sure input items are unique by the specified key.
+
+    The 'by' function must return a hashable key (the uniqueness key).
+
+    If item A and item B have the same key, then
+        if dedupe is true (the default) and A == B, B is discarded
+        if dedupe is false or A != B, raise an error
+    """
+    by_key: dict[tuple[str, ...], Any] = {}
+    for item in items:
+        multi_key = by(item)
+        found = False
+        for mkey in by_key:
+            for key in mkey:
+                if key in multi_key:
+                    found = True
+                    break
+            if found:
+                break
+        else:
+            by_key[tuple(multi_key)] = item
```

Comment:
I don't understand behavior of this function, why it's comparing sub keys?

It doesn't have unittest.

Reply:
So Idea is when you have components with more purls, it should take only first matching. But now I'm thinking about it maybe it needs more logic